torch/csrc/jit/passes/mkldnn_rewrite.cpp

```diff
@@ -0,0 +1,276 @@
+#include <torch/csrc/jit/passes/fold_conv_bn.h>
+#include <torch/csrc/jit/passes/freeze_module.h>
+#include <torch/csrc/jit/passes/graph_rewrite_helper.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/mkldnn_rewrite.h>
+#include <torch/csrc/jit/passes/remove_dropout.h>
+#include <torch/csrc/jit/runtime/graph_executor_impl.h>
+#include <torch/csrc/jit/tensorexpr/kernel.h>
+
+namespace torch {
+namespace jit {
+
+namespace mkldnn {
+namespace detail {
+
+#if AT_MKLDNN_ENABLED()
+bool mkldnn_fuser_enabled = true;
+#else
+bool mkldnn_fuser_enabled = false;
+#endif // AT_MKLDNN_ENABLED()
+
+} // namespace detail
+} // namespace mkldnn
+
+#if AT_MKLDNN_ENABLED()
+
+c10::VaryingShape<int64_t> getSizesOf(Node* n, size_t idx) {
+  auto tt = n->input(idx)->type()->cast<TensorType>();
+  return tt->sizes();
+}
+
+bool isContiguous(Value* v) {
+  auto const& tt = v->type()->cast<TensorType>();
+  if (!tt) {
+    return false;
+  }
+  if (!tt->isComplete()) {
+    return false;
+  }
+  auto const& sizes = tt->sizes().concrete_sizes();
+  auto const& strides = tt->strides().concrete_sizes();
+  if (!sizes || !strides) {
+    return false;
+  }
+  int ndims = (*sizes).size();
+  // TODO: only support 4D input for now
+  if (ndims != 4) {
+    return false;
+  }
+
+  // TODO: currently only support contiguous input. Channels last contiguous
+  // will be supported as a next step
+  return *strides == TensorType::contiguousStridesOf(*sizes);
+}
+
+void insertPrePackedConvOpForNode(Node* n) {
+  constexpr int POS_INPUT = 0;
+  constexpr int POS_WEIGHT = 1;
+  if (!isContiguous(n->input(POS_INPUT))) {
+    return;
+  }
+
+  if (!isContiguous(n->input(POS_WEIGHT))) {
+    return;
+  }
+
+  // Leave depthwise conv2d to NNC
+  if (tensorexpr::conv2dIsSupportedJit(n)) {
+    return;
+  }
+
+  WithInsertPoint guard(n);
+  auto graph = n->owningGraph();
+
+  auto input_sizes = getSizesOf(n, POS_INPUT);
+  IValue input_size_value(*input_sizes.concrete_sizes());
+  auto input_size = graph->insertConstant(input_size_value);
+
+  auto prepack_node = graph->create(
+      Symbol::fromQualString("mkldnn_prepacked::conv2d_prepack"), 1);
+
+  // skip input value
+  for (auto i = 1; i < n->inputs().size(); i++) {
+    Value* v = n->input(i);
+    prepack_node->addInput(v);
+  }
+  prepack_node->addInput(input_size);
+  auto attr = graph->insertConstant(IValue("none"));
+  prepack_node->addInput(attr);
+  prepack_node->output()->setType(
+      getCustomClass("__torch__.torch.classes.mkldnn.ConvOpContext"));
+  graph->insertNode(prepack_node);
+
+  auto prepack_conv = graph->insertNode(
+      graph->create(Symbol::fromQualString("mkldnn_prepacked::conv2d_run"), 1));
+  prepack_conv->addInput(n->input(0));
+  prepack_conv->addInput(prepack_node->output());
+  prepack_conv->output()->setType(n->output()->type()->cast<TensorType>());
+
+  n->output()->replaceAllUsesWith(prepack_conv->output());
+}
+
+bool canFuseOnDevice(Value* v) {
```

Comment:
Similarly, please try using existing functions from TE fuser when possible.

Reply:
Below are some differences compared with the existing `canFuseOnDevice` function:
https://github.com/pytorch/pytorch/blob/192aa3ad5f770214bd4ad26297deff08c88b239f/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L845-L867

1. Currently mkldnn prepacked conv2d only supports CPU
2. mkldnn prepacked conv2d requires MKLDNN to be enabled

Would it be okay that we add `canFuseOnDevice` in `mkldnn_rewrite.cpp` due to the above differences?

Reply:
So, all we do here now is checking that for tensor values the device is cpu and mkldnn is enabled. It is confusing to reuse the same name as the existing function `isFusableOnDevice` + `canFuseOnDevice` as these functions are intended to be used in a generic fuser that handles multiple devices + multiple fusible ops. I suggest renaming this function to something like "isTensorTypeCPU" and replace several functions with one. For the second part of the check - whether mkldnn is enabled - I suggest moving this check to the beginning of the pass and not check it on every node.

Reply:
Changed the function according to the comment.